src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs

```diff
@@ -37,18 +37,72 @@ public void Dispose()
             public override Stream CreateStream(DiagnosticBag diagnostics)
             {
                 Debug.Assert(_streamToDispose == null);
+                bool renameAttempted = false;
 
-                try
+                while (true)
                 {
-                    return _streamToDispose = _compiler.FileOpen(_filePath, PortableShim.FileMode.Create, PortableShim.FileAccess.ReadWrite, PortableShim.FileShare.None);
-                }
-                catch (Exception e)
-                {
-                    var messageProvider = _compiler.MessageProvider;
-                    diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_CantOpenFileWrite, Location.None, _filePath, e.Message));
-                    return null;
+                    try
+                    {
+                        return _streamToDispose = _compiler.FileOpen(_filePath, PortableShim.FileMode.Create, PortableShim.FileAccess.ReadWrite, PortableShim.FileShare.None);
+                    }
+                    catch (IOException e) when (!renameAttempted)
+                    {
+                        // Other process is reading the file preventing us to write to it.
+                        // We attempt to rename and delete the file in case the reader opened it with FileShare.Delete flag that
+                        // allows the file to be deleted by other processes.
+                        //
+                        // Note that if the file is marked "readonly" or the current user doesn't have sufficient privileges
+                        // the exception thrown is UnauthorizedAccessException, not IOException, so we won't attempt to delete the file.
+
+                        renameAttempted = true;
+                        try
+                        {
+                            if (PathUtilities.IsUnixLikePlatform)
+                            {
+                                // Unix & Mac are simple: just delete the file in the directory. 
+                                // The memory mapped content remains available for the reader.
+                                PortableShim.File.Delete(_filePath);
+                                continue;
+                            }
+
+                            const int eWin32SharingViolation = unchecked((int)0x80070020);
```

Comment:
@Pilchie Found the HResult in a list on msdn -- odd, neither google nor bing nor msdn search found it when I looked for SHARING_VIOLATION before. 
https://msdn.microsoft.com/en-us/library/windows/desktop/ms681382(v=vs.85).aspx 

```
ERROR_SHARING_VIOLATION
32 (0x20)
The process cannot access the file because it is being used by another process.
```